### PR TITLE
Feat/280/toggles

### LINKED ISSUE #280: [FEAT] - Make Thanks and Levels togglable
# Feature Request

## Is your feature request related to a problem? Please describe:

The `thanks` listener and `levels` listener are "always on". This can create a negative experience for some users/servers.

## Describe the solution you'd like

These should be modified to a server-based configuration (perhaps `true` for on and `false` for off). They can default to "on", but server config should allow a server to turn them off. 

> Note: The mention listener is called by the thanks listener - this approach will need to be changed as the mention listener should *always* be on.+

### PR DESCRIPTION
**Pull request by nhcarrigan**
# Pull Request

## Description:

This was a big one! 
* Creates a new model for toggles, to allow for boolean values instead of string values
* Adds a `toggle` option to the `config` command.
* Allows the `thanks` and `levels` listeners to be toggled.

## Related Issue:

Closes #280 

## Scope:

Did you remember to update the `package.json` version number?

- [ ] Major Version Update: X.\_.\_ (for complete code refactors, or large PRs that adjust most of the codebase)
- [ ] Minor Version Update: \_.X.\_ (for the addition of new commands)
- [ ] Patch Version Update: \_.\_.X (for bug fixes _in the code_.)
- [ ] No Version Update: \_.\_.\_ (no version update for additions to tests, documentation, or anything that isn't end-user facing.)

## Code Coverage:

Are all the unit tests passing and have you added tests that cover your changes? Run `npm run coverage`

Coverage Checklist:

- [X] All unit tests passing
- [ ] Code that my PR modifies has tests.
- [ ] PR defines validation steps.

## Documentation:

For _any_ version updates, please verify if the [documentation page](https://www.nhcarrigan.com/BeccaBot-documentation) needs an update. If it does, please [create an issue there](https://github.com/nhcarrigan/BeccaBot-documentation/issues/new?assignees=nhcarrigan&labels=%F0%9F%9A%A6+status%3A+awaiting+triage&template=update.md&title=%5BUPDATE%5D) to ensure it is not forgotten.

- [ ] My contribution does NOT require a documentation update.
- [X] My contribution DOES require a documentation update.

### CODE REVIEW

File: src/commands/server/config.ts [hunk 1]
```diff
@@ -63,7 +66,32 @@ const config: CommandInt = {
...
+          if (setType === "levels") {
+            const shouldLevel = await getToggleFromSettings("levels", guild);
+            await setToggle(guild.id, "levels", !shouldLevel);
+            await message.reply(
+              !shouldLevel
+                ? "I have turned the level feature on."
+                : "I have turned the level feature off."
+            );
+            return;
+          }
```
Comment (danielsolartech) [inline src/commands/server/config.ts:89]: You can combine this two cases:

```suggestion
          if (setType === "thanks" || setType === "levels") {
            const toggleSetting = await getToggleFromSettings(setType, guild);
            await setToggle(guild.id, setType, !toggleSetting);
            await message.reply(
              `I have turned the ${setType} feature ${!toggleSetting ? 'on' : 'off'}`
            );
            renturn;
          }
```

File: src/commands/server/config.ts [hunk 2]
```diff
@@ -378,6 +406,29 @@ const config: CommandInt = {
...
+
+      // Get the thanks setting from the database
+      const shouldThank = await getToggleFromSettings("thanks", guild);
+
+      // Add the thanks setting to an embed field.
+      configEmbed.addField(
+        "Thanks",
+        shouldThank
+          ? "I will congratulate users when another user thanks them."
+          : "I will NOT congratulate users when another user thanks them."
```
Comment (danielsolartech) [inline src/commands/server/config.ts:418]:
```suggestion
        `I will${shouldThank ? '' : ' NOT'} congratule users when another user thanks them`
```

File: src/commands/server/config.ts [hunk 3]
```diff
@@ -378,6 +406,29 @@ const config: CommandInt = {
...
+
+      //get the levels setting from the database
+      const shouldLevel = await getToggleFromSettings("levels", guild);
+
+      // Add the levels setting to an embed field.
+      configEmbed.addField(
+        "Levels",
+        shouldLevel
+          ? "I will give users experience points for being active"
+          : "I will NOT give users experience points for being active."
```
Comment (danielsolartech) [inline src/commands/server/config.ts:429]:
```suggestion
        `I will${shouldLevel ? '' : ' NOT'} give users experience points for being active.`
```

File: src/utils/extendsClientToClientInt.ts [hunk 4]
```diff
@@ -93,7 +118,7 @@ async function getRoleFromSettings(
   // Check if the role exists in the database.
   if (roleSetting) {
     // Get the role from the server roles.
-    const role = guild.roles.cache.get(roleSetting.value);
+    const role = guild.roles.cache.get(roleSetting.value as string);
```
[the remark this answers is not in the document]
Comment (nhcarrigan) [reply src/utils/extendsClientToClientInt.ts:121]: Oops, that one was left over from my attempt with the `SettingModel`.

File: src/listeners/thanksListener.ts [hunk 5]
```diff
@@ -1,12 +1,29 @@
...
 
 const thanksListener: ListenerInt = {
   name: "Thanks",
   description: "Congratulates users who are thanked.",
   run: async (message) => {
     try {
-      const { author, bot, channel } = message;
+      const { author, bot, channel, guild } = message;
+
+      if (!guild) {
```
[the remark this answers is not in the document]
Comment (nhcarrigan) [reply src/listeners/thanksListener.ts:12]: When does a guild not exist? 

It's in the case of a DM, right? In which case, the bot has a different handler for DMs so this should be okay.

[the remark this answers is not in the document]
Comment (nhcarrigan) [reply src/listeners/thanksListener.ts:12]: Additionally, the `botMentionListener` handles the logic for whether the bot is mentioned - this just ensures that listener only gets called if the `thanksListener` fails.

**Review by danielsolartech — APPROVED** [review]

**Pull request merged**